### PR TITLE
[onert] Correct the shape when _keep_dims is false in backward of MeanLayer

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit corrects the shape by creating a temporary shape having the same rank as the input when _keep_dims is false because MeanGrad does not support other rank cases.

- Revert https://github.com/Samsung/ONE/pull/13351
- Fix `GenModelTrain.NonTrainableOps_FC_Mean` test case

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

/cc @ragmani

### CODE REVIEW

**Review by hseok-oh — APPROVED** [review]
LGTM

**Comment by hseok-oh** [comment]
@nnfw-bot test onert-x64-release

**Comment by glistening** [comment]
I cannot find the change for
- Fix GenModelTrain.NonTrainableOps_FC_Mean test case

in this PR.

It seems to only revert
- https://github.com/Samsung/ONE/pull/13351

Then, I wonder how CI passes `GenModelTrain.NonTrainableOps_FC_Mean`.

https://github.com/Samsung/ONE/pull/13394#issuecomment-2219849675 is the result of filter-out `
GenModelTrain.NonTrainableOps_FC_Mean`.

cc @ragmani

File: runtime/onert/backend/train/ops/MeanLayer.cc [hunk 1]
```diff
@@ -48,11 +48,29 @@ void MeanLayer::forward(bool) { cpu::ops::MeanLayer::run(); }
...
+    keep_dim_shape.ReplaceWith(getShape(_input));
+    auto axes_vec = cpu::ops::getReducerAxes(_axes);
+    for (const auto &axis : axes_vec)
+    {
+      keep_dim_shape.SetDim(axis, 1);
+    }
+  }
+  else
+  {
+    keep_dim_shape.ReplaceWith(getShape(_back_prop_output));
```
Comment (hseok-oh) [inline runtime/onert/backend/train/ops/MeanLayer.cc:66]: @glistening This line is updated: `_back_prop_input` -> `_back_prop_output`

**Comment by jyoungyun** [comment]
@glistening 

> https://github.com/Samsung/ONE/pull/13394#issuecomment-2219849675 is the result of filter-out `
GenModelTrain.NonTrainableOps_FC_Mean`.

```bash
$ ./Product/out/unittest/nnfw_api_gtest --gtest_filter=GenModelTrain.NonTrainableOps_FC_Mean
Note: Google Test filter = GenModelTrain.NonTrainableOps_FC_Mean
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from GenModelTrain
[ RUN      ] GenModelTrain.NonTrainableOps_FC_Mean
[       OK ] GenModelTrain.NonTrainableOps_FC_Mean (1 ms)
[----------] 1 test from GenModelTrain (2 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2 ms total)
[  PASSED  ] 1 test.
```

Here is the `GenModelTrain.NonTrainableOps_FC_Mean` test result.

**Comment by ragmani** [comment]
I have 2 things that I don't understand about this PR.
1. AFAIK, back-propagated shapes are always the same with forwarded shapes. I don't understand why the shape can be differrent only in mean op. Why are the nodes that will be executed after the mean op not affected?

2. Why only change the shape, but not the size of the buffer?

**Comment by jyoungyun** [comment]
@ragmani 

> 1. AFAIK, back-propagated shapes are always the same with forwarded shapes. I don't understand why the shape can be differrent only in mean op. Why are the nodes that will be executed after the mean op not affected?

In the MeanGrad implementation, the `BroadcastTo` functionis used. This code only changes the shape to recognize the `BroadcastTo` function.

Here is an example of this case:

```
backprop_input shape: 10,7,7,1280
backprop_output shape: 10, 1280
```

If the backprop_output shape passes directly through the MeanGrad function when `_keep_dims` is false, the `BroadcastTo` function does not work because the rank does not match. This code corrects the shape so that the `BroadcastTo` can recognize it.

keep_dim_shape: 10,1,1,1280

If `_keep_dims` is false, keep_dim shape is used instead of backprop_output shape.

> 2. Why only change the shape, but not the size of the buffer?

It does not change the buffer size, it only changes the shape of the buffer.

If you do not agree this approach, MeanGrad implementation should be modified.

Thank you for code review during the trip. :)

**Review by glistening — APPROVED** [review]
LGTM

**Comment by ragmani** [comment]
@jyoungyun Thanks for your kindly explanation. But I'm still not sure if the conditional shape change is needed there.
Keep dims option is applyed in forwarding. The shapes of backwarding and forwarding are the same. If so, I guess backprop_output is already been applied keep dim option. Could you check it?

**Comment by jyoungyun** [comment]
@ragmani 

> The shapes of backwarding and forwarding are the same. If so, I guess backprop_output is already been applied keep dim option. Could you check it?

This issue is caused by using the `BroadcastTo` function in `MeanGrad` implementation. The `BroadcastTo` has restrictions on shape and these changes are intended to meet the restrictions. Therefore, this issue is not happened in forwarding.

**Comment by ragmani** [comment]
@jyoungyun So, could you add a test that fails by `Broadcast` restrictions without this PR to prevent wrong changes I did later?

**Pull request merged**